source/exe/hot_restart.h

```diff
@@ -138,8 +138,12 @@ class HotRestartImpl : public HotRestart,
   };
 
   struct RpcBase {
-    RpcBase(RpcMessageType type, uint64_t length = sizeof(RpcBase))
-        : type_(type), length_(length) {}
+    RpcBase(RpcMessageType type, uint64_t length = sizeof(RpcBase)) {
+      static_assert(std::is_standard_layout<RpcBase>::value, "is_standard_layout<RpcBase>");
```

Comment:
technically I think you actually want to check that the derived class is standard layout since you are about to memset beyond the length of RpcBase. Perhaps just do the check for one of the derived classes? Might be worth a small comment to in here?